### PR TITLE
[21.05] sway: 1.6 -> 1.6.1, wlroots: 0.13.0 -> 0.14.0

### LINKED ISSUE #127413: sway: regular segfaults
**Describe the bug**
I'm regularly observing `sway` crashes, especially after unlocking with `swaylock`. This happens all two to three days. I get the following backtrace when investigating the crash via `coredumpctl gdb`:

```
(gdb) bt
#0  0x00007fa47c0289b0 in malloc_consolidate () from /nix/store/ikl21vjfq900ccbqg1xasp83kadw6q8y-glibc-2.32-46/lib/libc.so.6
#1  0x00007fa47c02ae3d in _int_malloc () from /nix/store/ikl21vjfq900ccbqg1xasp83kadw6q8y-glibc-2.32-46/lib/libc.so.6
#2  0x00007fa47c02cb98 in malloc () from /nix/store/ikl21vjfq900ccbqg1xasp83kadw6q8y-glibc-2.32-46/lib/libc.so.6
#3  0x00007fa47b8030cc in drmModeAtomicAddProperty () from /nix/store/fj65xn3jm6fzyp5anki4r2v0c351dmid-libdrm-2.4.105/lib/libdrm.so.2
#4  0x00007fa47c21b9ac in atomic_add () from /nix/store/zkn8bk91glf1gki5lw6ch67fwdx807p6-wlroots-0.13.0/lib/libwlroots.so.8
#5  0x00007fa47c21bca3 in atomic_crtc_commit () from /nix/store/zkn8bk91glf1gki5lw6ch67fwdx807p6-wlroots-0.13.0/lib/libwlroots.so.8
#6  0x00007fa47c21d0a2 in drm_crtc_commit () from /nix/store/zkn8bk91glf1gki5lw6ch67fwdx807p6-wlroots-0.13.0/lib/libwlroots.so.8
#7  0x00007fa47c21de16 in drm_crtc_page_flip () from /nix/store/zkn8bk91glf1gki5lw6ch67fwdx807p6-wlroots-0.13.0/lib/libwlroots.so.8
#8  0x00007fa47c252839 in wlr_output_commit () from /nix/store/zkn8bk91glf1gki5lw6ch67fwdx807p6-wlroots-0.13.0/lib/libwlroots.so.8
#9  0x0000000000421e8c in output_render ()
#10 0x000000000041e9c0 in output_repaint_timer_handler ()
#11 0x000000000041ebe1 in damage_handle_frame ()
#12 0x00007fa47c2613fc in wlr_signal_emit_safe () from /nix/store/zkn8bk91glf1gki5lw6ch67fwdx807p6-wlroots-0.13.0/lib/libwlroots.so.8
#13 0x00007fa47c2613fc in wlr_signal_emit_safe () from /nix/store/zkn8bk91glf1gki5lw6ch67fwdx807p6-wlroots-0.13.0/lib/libwlroots.so.8
#14 0x00007fa47c2b33bb in wl_event_loop_dispatch_idle () from /nix/store/28n6s73hff7b7mcm09x5pm4qk86w0snv-wayland-1.19.0/lib/libwayland-server.so.0
#15 0x00007fa47c2b34f5 in wl_event_loop_dispatch () from /nix/store/28n6s73hff7b7mcm09x5pm4qk86w0snv-wayland-1.19.0/lib/libwayland-server.so.0
#16 0x00007fa47c2b1135 in wl_display_run () from /nix/store/28n6s73hff7b7mcm09x5pm4qk86w0snv-wayland-1.19.0/lib/libwayland-server.so.0
#17 0x000000000040faae in main ()
```

**To Reproduce**
Unfortunately this happens pretty unregularly and I haven't identified a pattern yet.

**Expected behavior**
No crashes when unlocking `sway`.

**Additional context**
Could be related to https://github.com/swaywm/wlroots/issues/2475 or https://github.com/swaywm/wlroots/issues/204, but both seem to have been fixed in the past.

**Notify maintainers**
@primeos @Synthetica9 

**Metadata**
```
λ ma27 [~] → nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.43, NixOS, 21.05.20210612.589a8bb (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210601_5985b8b`
 - channels(ma27): `""`
 - nixpkgs: `/nix/store/rsp0183gda98drf021q9325154q0sbby-source`
```

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

Backport PR for #128036, #127897 and #123519 (not so sure about the last one, I decided to update it as well to hopefully fix #127413 locally).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by primeos** [comment]
Not sure if this is really a good idea. Maybe we should wait a bit to see if anyone on nixos-unstable experiences any regressions with wlroots 0.14? I also thought you wanted to test this for ~10 days first (https://github.com/NixOS/nixpkgs/issues/127413#issuecomment-869134761) - or was that just for that issue and not the backport? Additionally I though we intended to backport the wlroots update as `wlroots_0_14` (https://github.com/NixOS/nixpkgs/pull/128036#issuecomment-868341743) or are you against that idea?

The `libdrm` update is also a bit scary (lots of rebuilds) but at least I've never heard of any `libdrm` regressions since I started maintaining it a few releases ago or so.

Anyway, just some things to consider, this isn't a hard objection / veto.

PS: Ideally we should also state the reason: "4. When the reason to backport is not obvious from the original commit message, use git cherry-pick -xe <original commit> and add a reason."

**Comment by Ma27** [comment]
> I also thought you wanted to test this for ~10 days first (#127413 (comment)) - or was that just for that issue and not the backport? Additionally I though we intended to backport the wlroots update as wlroots_0_14 (#128036 (comment)) or are you against that idea?


Absolutely! Just thought I'd open it now to give other people on 21.05 the same opportunity. To make that clear, I decided to mark it as draft. (Note: I expected it to be a bit more work to backport this which is another reason why this branch exists :))

> The libdrm update is also a bit scary (lots of rebuilds) but at least I've never heard of any libdrm regressions since I started maintaining it a few releases ago or so.

My thinking was kinda simple (and perhaps dumb) here: the crash I've seen happened in a communication between sway and libdrm, so updating it to the latest version was something I wanted to try out and it works fine for me.

But as I said in the original post, I'm not so sure if that's a good idea, so we can leave it out, of course.

**Comment by Ma27** [comment]
I removed the `libdrm`-update since it hasn't solved my problem (I'm currently trying out various WLR_* settings). `sway` with a newer wlroots worked fine regardless.

**Comment by Ma27** [comment]
The (more, but currently less frequent) crashes I see were already an issue on sway 1.6, apart from that I don't see any issues with sway 1.6.1/wlroots 0.14.0 on 21.05. Because of that I'm going to mark it as ready for review now.

File: pkgs/applications/window-managers/cagebreak/default.nix [hunk 1]
```diff
@@ -31,6 +32,15 @@ stdenv.mkDerivation rec {
     hash = "sha256-1IztedN5/I/4TDKHLJ26fSrDsvJ5QAr+cbzS2PQITDE=";
   };
 
+  patches = [
+    # To fix the build with wlroots 0.14.0:
+    (fetchpatch {
+      # Add fixes for wlroots 0.14.0
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/cagebreak/default.nix:38]: Redundant comment.

File: pkgs/applications/window-managers/labwc/default.nix [hunk 2]
```diff
@@ -30,6 +31,15 @@ stdenv.mkDerivation rec {
     sha256 = "0rhniv5j4bypqxxj0nbpa3hclmn8znal9rldv0mrgbizn3wsbs54";
   };
 
+  patches = [
+    # To fix the build with wlroots 0.14:
+    (fetchpatch {
+      # output: access texture width/height directly
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/labwc/default.nix:37]: Redundant?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by Ma27** [comment]
> Redundant comment.

Yeah, I really don't want to change commits during a backport though unless it's strictly needed.

**Review by primeos — CHANGES_REQUESTED** [review]
This PR currently contains both `libdrm: 2.4.105 -> 2.4.106` and a commit which reverts that update so I'd say we should certainly drop both (even though it doesn't really change the end result).

And what's the current state on this?:
> Additionally I though we intended to backport the wlroots update as wlroots_0_14 (#128036 (comment)) or are you against that idea?

Currently this PR doesn't use that approach. If we don't want to use that approach it might make sense to drop wlroots 0.13 entirely like on `nixos-unstable`: https://github.com/NixOS/nixpkgs/issues/118838

Anyway, I'm fine with both approaches so this is mainly a question because no explicit decision was made so far (apart from the diff).

> Edit: before anybody considers a merge here, an ACK from @primeos should be awaited :)

I'll give this an ACK then (despite requesting changes here) in case anyone wants to merge it ;) I haven't done any testing and don't use NixOS 21.05 but the idea and diff LGTM (although certainly not risk free, there are a few changes that could be considered "breaking" but hopefully it won't affect any actual setups in a meaningful way). Other than the two libdrm commits this should already be ok.

**Comment by primeos** [comment]
https://octodon.social/@emersion/106544587943751964
> Alright, I think I fixed all known wlroots 0.14.0 regressions. Some were pretty tricky. 0.14.1 coming soon!

Just as an FYI (not sure if it's worth it to wait for wlroots 0.14.1 before merging this or not).

**Comment by primeos** [comment]
@Ma27 if you're interested we could/should also backport #129667 (should be fine to go directly into `staging-next-21.05` so that it lands together with this PR once #129778 is merged).

**Pull request merged**